### PR TITLE
Improvements for Menu resolver

### PR DESCRIPTION
**Pull request by mihailspopovs4**
Simplified code that returns Menu.

### CODE REVIEW

File: src/Model/Resolver/Menu.php [hunk 1]
```diff
@@ -84,43 +71,56 @@ public function resolve(
...
+
+        $menuItems = $this->itemCollectionFactory
+            ->create()
+            ->addMenuFilter($menuId)
+            ->addStatusFilter()
+            ->setParentIdOrder()
+            ->setPositionOrder()
+            ->getData();
+
+        foreach ($menuItems as &$item) {
```
Comment (eli-l) [inline src/Model/Resolver/Menu.php]: Let's use array_map for more verbose code

File: src/Model/Resolver/Menu.php [hunk 2]
```diff
@@ -84,43 +71,56 @@ public function resolve(
...
-                };
-            }
+    /**
+     * @param string $menuId
+     * @return array
+     * @throws \Magento\Framework\Exception\NoSuchEntityException
+     */
+    private function getMenuItems(string $menuId): array
+    {
+        $categoryIdKey = 'category_id';
```
Comment (eli-l) [inline src/Model/Resolver/Menu.php]: It would be better to define as a constant

File: src/Model/Resolver/Menu.php [hunk 3]
```diff
@@ -84,43 +71,56 @@ public function resolve(
...
-
-            foreach ($menuData['items'] as &$item) {
-                if (isset($item['category_id'])) {
-                    $category = $this->categoryRepository->get($item['category_id']);
-                    $item['url'] = sprintf('/%s', $category->getUrlPath());
-                }
-            }
+    )
+    {
+        $identifier = $args['identifier'];
```
Comment (eli-l) [inline src/Model/Resolver/Menu.php:88]: Please use ResourceModel instead of collection.

**Review by eli-l — CHANGES_REQUESTED** [review]

File: src/Model/Resolver/Menu.php [hunk 4]
```diff
@@ -84,43 +75,49 @@ public function resolve(
...
-            /** Updated with identifier filtering */
-            $menu->addFieldToFilter('identifier', $args['identifier'])->load();
+    )
+    {
+        $identifier = $args['identifier'];
 
-            $menuData = $menu->getFirstItem()->getData();
+        $menu = $this->menuFactory
+            ->create()
+            ->load($identifier);
```
Comment (eli-l) [inline src/Model/Resolver/Menu.php]: Do not load on Model, use Resource model, ie:
$menu = $this->menuFactory->create();
$this->resourceModel->load($menu, $identifier)

File: src/Model/Resolver/Menu.php [hunk 5]
```diff
@@ -84,43 +75,49 @@ public function resolve(
...
+     */
+    private function getMenuItems(string $menuId): array
+    {
+        $menuItems = $this->itemCollectionFactory
+            ->create()
+            ->addMenuFilter($menuId)
+            ->addStatusFilter()
+            ->setParentIdOrder()
+            ->setPositionOrder()
+            ->getData();
```
Comment (eli-l) [inline src/Model/Resolver/Menu.php:116]: ->getData() is unnecessary

File: src/Model/Resolver/Menu.php [hunk 6]
```diff
@@ -28,15 +25,12 @@
  */
 class Menu implements ResolverInterface
 {
-    /**
-     * @var ValueFactory
-     */
-    private $valueFactory;
+    const CATEGORY_ID_KEY = 'category_id';
```
Comment (eli-l) [inline src/Model/Resolver/Menu.php]: In latest PHP versions there are scoped constants, but usually these are public, however we still must declare the scope: **public const CATEGORY_ID_KEY=...**

File: src/Model/Resolver/Menu.php [hunk 7]
```diff
@@ -84,43 +75,49 @@ public function resolve(
...
+            ->addMenuFilter($menuId)
+            ->addStatusFilter()
+            ->setParentIdOrder()
+            ->setPositionOrder()
+            ->getData();
+
+        return array_map(function ($item) {
+            if (isset($item[static::CATEGORY_ID_KEY])) {
+                $categoryUrlPath = $this->categoryRepository
+                    ->get($item[static::CATEGORY_ID_KEY])
```
Comment (eli-l) [inline src/Model/Resolver/Menu.php]: static is used for late static binding that is not needed here. Use self:: instead.
Here are the docs on late static binding: https://www.php.net/manual/en/language.oop5.late-static-bindings.php

**Review by eli-l — CHANGES_REQUESTED** [review]

File: src/Model/Resolver/Menu.php (hunk 5 above)
Comment (mihailspopovs4) [reply src/Model/Resolver/Menu.php:116]: I have tried to remove `->getData` call before, but it throws exception saying that `array` must be provided for `array_map`. It seems it does not iterate it like `foreach`.

File: src/Model/Resolver/Menu.php [hunk 8]
```diff
@@ -84,43 +83,48 @@ public function resolve(
...
-
-            /** Updated with menu id taken from menu data */
-            $items->addMenuFilter($menuData['menu_id'])
-                ->addStatusFilter()
-                ->setParentIdOrder()
-                ->setPositionOrder();
-
-            $menuData['items'] = $items->getData();
+        if ($menu->getId() === null) {
+            throw new \InvalidArgumentException("Could not find menu with identifier '${identifier}'");
```
Comment (eli-l) [inline src/Model/Resolver/Menu.php]: No need for '${}' wrappers, double-quotes allows to use variables directly, without templating, placeholders etc, i.e. "Identifier: $identifier".
However, a better way is to handle string templates using sprintf() function.

File: src/Model/Resolver/Menu.php [hunk 9]
```diff
@@ -84,43 +83,48 @@ public function resolve(
...
-
-            /** Updated with identifier filtering */
-            $menu->addFieldToFilter('identifier', $args['identifier'])->load();
+    )
+    {
+        $identifier = $args['identifier'];
 
-            $menuData = $menu->getFirstItem()->getData();
+        $menu = $this->menuFactory->create();
+        $this->menuResourceModel->load($menu, $identifier);
```
Comment (eli-l) [inline src/Model/Resolver/Menu.php:91]: We could pass $args['identifier'] directly.

**Review by eli-l — CHANGES_REQUESTED** [review]

File: src/Model/Resolver/Menu.php (hunk 8 above)
Comment (mihailspopovs4) [reply src/Model/Resolver/Menu.php]: Changed code to user `sprintf`. Also did for menu item url creation.

File: src/Model/Resolver/Menu.php (hunk 9 above)
Comment (mihailspopovs4) [reply src/Model/Resolver/Menu.php:91]: We use `$identifier` variable later on, when throwing exception.

**Review by eli-l — APPROVED** [review]

**Pull request merged**